pkg/remoteresolution/resolver/bundle/resolver.go

```diff
@@ -1,87 +1,144 @@
 	}
 }
 
-// Validate ensures reqolution request spec from a request are as expected.
+// GetConfigName returns the name of the bundle resolver's configmap.
+func (r *Resolver) GetConfigName(_ context.Context) string {
+	return bundleresolution.ConfigMapName
+}
+
+// Validate ensures parameters from a request are as expected.
 func (r *Resolver) Validate(ctx context.Context, req *v1beta1.ResolutionRequestSpec) error {
-	if len(req.Params) > 0 {
-		return bundle.ValidateParams(ctx, req.Params)
+	return bundleresolution.ValidateParams(ctx, req.Params)
+}
+
+// IsImmutable implements ImmutabilityChecker.IsImmutable
+// Returns true if the bundle parameter contains a digest reference (@sha256:...)
+func (r *Resolver) IsImmutable(params []v1.Param) bool {
+	var bundleRef string
+	for _, param := range params {
+		if param.Name == bundleresolution.ParamBundle {
+			bundleRef = param.Value.StringVal
+			break
+		}
+	}
+
+	// Checks if the given string looks like an OCI pull spec by digest.
+	// A digest is typically in the format of @sha256:<hash> or :<tag>@sha256:<hash>
+	// Check for @sha256: pattern
+	if strings.Contains(bundleRef, "@sha256:") {
+		return true
 	}
-	// Remove this error once validate url has been implemented.
-	return errors.New("cannot validate request. the Validate method has not been implemented.")
+	// TODO(twoGiants): doesn't look right => write test => should check for two colons
+	// Check for :<tag>@sha256: pattern
+	if strings.Contains(bundleRef, ":") && strings.Contains(bundleRef, "@sha256:") {
+		return true
+	}
+
+	return false
 }
 
-// Resolve uses the given request spec resolve the requested file or resource.
+// Resolve uses the given params to resolve the requested file or resource.
 func (r *Resolver) Resolve(ctx context.Context, req *v1beta1.ResolutionRequestSpec) (resolutionframework.ResolvedResource, error) {
-	if len(req.Params) > 0 {
-		return bundle.ResolveRequest(ctx, r.kubeClientSet, req)
+	if len(req.Params) == 0 {
+		return nil, errors.New("no params")
+	}
+
+	// Verify resolver was initialized - kubeClientSet must be set by Initialize()
+	if r.kubeClientSet == nil {
+		return nil, errors.New("bundle resolver not properly initialized: Initialize() must be called before Resolve()")
+	}
+
+	// Defensive: set default resolveRequestFunc if somehow nil
+	resolveFunc := r.resolveRequestFunc
+	if resolveFunc == nil {
+		resolveFunc = bundleresolution.ResolveRequest
 	}
-	// Remove this error once resolution of url has been implemented.
-	return nil, errors.New("the Resolve method has not been implemented.")
+
+	if r.useCache(ctx, req) {
+		return framework.RunCommonCacheOperations(
+			ctx,
+			req.Params,
+			LabelValueBundleResolverType,
+			func() (resolutionframework.ResolvedResource, error) {
+				return resolveFunc(ctx, r.kubeClientSet, req)
+			},
+		)
+	}
+
+	return resolveFunc(ctx, r.kubeClientSet, req)
+}
+
+func (r *Resolver) useCache(ctx context.Context, req *v1beta1.ResolutionRequestSpec) bool {
```

Comment:
This is duplicated on each resolver and it's really just a very thin wrapper. I think we could just inline `framework.ShouldUseCache...` directly in the `if` above.